Panels/SongEditor/SongEditor.gd

```diff
@@ -1,14 +1,52 @@
 extends VBoxContainer
 
-var InstrumentTrack = preload("./InstrumentTrack.tscn")
+var track_name = preload("./TrackName.tscn")
+var song_segment = preload("./SongSegment.tscn")
+
+onready var names = $TracksScroll/HBox/Names
+onready var segments = $TracksScroll/HBox/SegmentScroll/VBoxContainer
+
+# styles
+var style_1 = preload("res://Theme/Default/SongButton1.tres")
+var style_2 = preload("res://Theme/Default/SongButton2.tres")
+var style_pressed = preload("res://Theme/Default/SongButtonPressed.tres")
 
 # Takes a Button since it conveniently sends an icon and message
+# TODO: Not use button as param
 func add_track(instrument: Button):
-	var inst = InstrumentTrack.instance()
-	inst.set_instrument(instrument.icon, instrument.text)
-	$ScrollContainer/Tracks.add_child(inst)
-#	inst.size_flags_horizontal |= SIZE_FILL
-#	inst.rect_min_size.x = 192
-
-func _on_InstrumentsPanel_instrument_chosen(instrument) -> void:
-	add_track(instrument)
+	var name = track_name.instance()
+	var segment_container = HBoxContainer.new()
+	name.set_instrument(instrument.icon, instrument.text, segment_container)
+	names.add_child(name)
+	if Global.segments == 0:
+		Global.segments = segments.get_parent().get_size().x/25
+	add_segments(segment_container, name.rect_size.y, true)
+	segments.add_child(segment_container)
+
+func add_segments(segment_container: HBoxContainer, size, style):
+	for i in range(segment_container.get_child_count(),
+	Global.segments):
+		var segment = Button.new()
+		segment.rect_min_size.x = 20
+		segment.rect_min_size.y = size
+		segment.focus_mode = Control.FOCUS_NONE
+		segment.add_stylebox_override("disabled", style_pressed)
+		if i % 4 == 0: style = !style
+		segment.add_stylebox_override("normal", style_1 if style else style_2)
+		segment.add_stylebox_override("hover", style_1 if style else style_2)
+		segment.connect("gui_input", self, "segment_input", [segment])
+		segment_container.add_child(segment)
+
+func segment_input(event, segment: Button):
+	if event is InputEventMouseButton:
+		if event.pressed:
+			if event.button_index == 1:
+				segment.disabled = true
+				if segment.get_index() == Global.segments - 1:
+					Global.segments += segments.get_parent().get_size().x/25
+					var style = segment.get_stylebox("normal") == style_1
+					add_segments(segment.get_parent(), segment.get_size().x, segment)
+#			if event.doubleclick :
+#				$Panel.visible = !$Panel.visible
+			if event.button_index == 3:
+				segment.disabled = false
```

Comment:
Might wanna collapse some of these if statements to avoid nesting them too many times

Reply:
Converted if's to a match and joined to ifs together